packager/packager.cc

```diff
@@ -0,0 +1,731 @@
+        return original;
+    }
+
+    ast::TreePtr preTransformAssign(core::MutableContext ctx, ast::TreePtr original) {
+        illegalNode(ctx, original->loc, "`=`");
+        return original;
+    }
+
+    ast::TreePtr preTransformHash(core::MutableContext ctx, ast::TreePtr original) {
+        illegalNode(ctx, original->loc, "hash literals");
+        return original;
+    }
+
+    ast::TreePtr preTransformArray(core::MutableContext ctx, ast::TreePtr original) {
+        illegalNode(ctx, original->loc, "array literals");
+        return original;
+    }
+
+    ast::TreePtr preTransformMethodDef(core::MutableContext ctx, ast::TreePtr original) {
+        illegalNode(ctx, original->loc, "method definitions");
+        return original;
+    }
+
+    ast::TreePtr preTransformBlock(core::MutableContext ctx, ast::TreePtr original) {
+        illegalNode(ctx, original->loc, "blocks");
+        return original;
+    }
+
+    ast::TreePtr preTransformInsSeq(core::MutableContext ctx, ast::TreePtr original) {
+        illegalNode(ctx, original->loc, "`begin` and `end`");
+        return original;
+    }
+};
+
+// Sanity checks package files, mutates arguments to export / export_methods to point to item in namespace,
+// builds up the expression injected into packages that import the package, and codegens the <PackagedMethods>  module.
+unique_ptr<PackageInfo> getPackageInfo(core::MutableContext ctx, ast::ParsedFile &package) {
+    ENFORCE(package.file.exists());
+    ENFORCE(package.file.data(ctx).sourceType == core::File::Type::Package);
+    // Assumption: Root of AST is <root> class.
+    ENFORCE(ast::isa_tree<ast::ClassDef>(package.tree));
+    ENFORCE(ast::cast_tree_nonnull<ast::ClassDef>(package.tree).symbol == core::Symbols::root());
+    auto packageFilePath = package.file.data(ctx).path();
+    ENFORCE(FileOps::getFileName(packageFilePath) == PACKAGE_FILE_NAME);
+    PackageInfoFinder finder;
+    package.tree = ast::TreeMap::apply(ctx, finder, move(package.tree));
+    finder.finalize(ctx);
+    if (finder.info) {
+        finder.info->packagePathPrefix = packageFilePath.substr(0, packageFilePath.find_last_of('/') + 1);
+    }
+    return move(finder.info);
+}
+
+// Add:
+//    module <PackageRegistry>::Mangled_Name_Package
+//      module ImportedPackage1
+//        # imported aliases go here
+//      end
+//    end
+// ...to __package.rb files to set up the package namespace.
+ast::ParsedFile rewritePackage(core::Context ctx, ast::ParsedFile file, const PackageDB &packageDB) {
+    ast::ClassDef::RHS_store importedPackages;
+
+    auto package = packageDB.getPackageByFile(file.file);
+    if (package == nullptr) {
+        // We already produced an error on this package when producing its package info.
+        // The correct course of action is to abort the transform.
+        return file;
+    }
+
+    // Sanity check: __package.rb files _must_ be typed: strict
+    if (file.file.data(ctx).strictLevel != core::StrictLevel::Strict) {
```

Comment:
```suggestion
    if (file.file.data(ctx).strictLevel < core::StrictLevel::Strict) {
```

There are other levels stricter than strict.